### DISCUSSION TITLE
Mapping visualization attributes to the correct node/edge

### DISCUSSION THREAD

**Discussion by rossbar** [networkx/networkx#5706] [General]
There can be mapping problems when specifying a sequence of visualization properties (e.g. a list of edge colors). This arises from the fact that the default `edgelist=list(G.edges)` might have a different ordering than a user expects. For example:

```python
>>> G = nx.Graph([(0, 1), (2, 1), (3, 4), (2, 4)])
>>> edge_colors = ["red", "green", "blue", "black"]
>>> pos = nx.circular_layout(G)
>>> nx.draw_networkx_nodes(G, pos)
>>> nx.draw_networkx_labels(G, pos)
>>> nx.draw_networkx_edges(G, pos, edge_color=edge_colors)
```

I naively expected the `edge_colors` to map to the edges in the order they were specified in the `Graph` constructor (i.e. (0, 1)->red, (2, 1)->green, (3, 4)->blue and (2, 4)->black). However, the edgecolors of the last two edges are flipped:

![ed](https://user-images.githubusercontent.com/1268991/155791887-b388cd0f-fc6c-4905-8b61-590887819090.png)

This is due to the fact that `list(G.edges)` returns a different ordering than the edgelist used to create the graph:

```
>>> list(G.edges)
[(0, 1), (1, 2), (2, 4), (3, 4)]
```

So what this means is - the only way you can currently *guarantee* that a sequence of properties (colors, sizes, etc.) maps to the edges you expect is to also pass in `edgelist` where the ordering explicitly matches the order of the properties! (though the self-loops ordering would likely still be broken, at least for undirected graphs).

Note: this is excerpted from a larger discussion originally in #5106

**Comment by dschult** [comment]
This is an important long term discussion point for our drawing tools. I am running into similar issues with the LaTeX interface.
The API that feels nicest to me is the one used by `G.edges`. That is, to store the drawing attributes in a dict keyed by edges to an attribute dict keyed by attribute name to attribute value.  It might be better for drawing to have a dict for each attribute keyed by edge to the value of that attribute.  But these are all dict oriented.  Matplotlib and other systems seem to be more array oriented where the order of the edges is assumed known and fixed and the attributes are arrays/lists of values.

If we continue with the attributes as arrays/lists, Maybe we should require that edgelist be specified.  Another approach would be to require drawing attributes to be stored in the graph.   Thoughts?

**Comment by mjschwenne** [comment]
I haven't need to visualize too many graphs through NetworkX, but for an assignment in my Artificial Intelligence class we needed to draw a state space diagram which I choose to generate in NetworkX and visualized used the pygraphviz interface and styling information about color, fill, etc, is stored as node and edge attributes respectively. 

Personally, I thought that that method of doing it was easy to use and concise since it removed all of this ambiguity about weather styling is applied correctly since it had to be explicitly encoded into the graph. Whenever I've used some of the other options like Matplotlib I've always been aware of the underlying assumption that the orders have to match and here is an example of were they didn't! Also, even though it is technically only a pass through to another library there is some level of precedent for it in NetworkX.  

On the other hand, I would understand the argument that it could clutter the graph if printing like all of the attributes on a node or edge after it is visualized but there are ways to mitigate that by later removing the attributes or annotating a copy of the graph for visualization, or only putting those attributes on the graph right before visualizing it, which is what I did.

**Comment by rossbar** [comment]
> Another approach would be to require drawing attributes to be stored in the graph. Thoughts?

I like the idea of *supporting* this but not necessarily *requiring* it, which would be a significant change to the interface. The problem here is that the `nx_pylab` interface is *already* quite complex and arguably fragile as a result. Adding *more* complexity on top of `nx_pylab` might only make things worse. Nevertheless the end-goal of having the ability to extract visualization properties directly from node and edge attributes as mentioned by @dschult and @mjschwenne is certainly appealing.

**Reply by dschult** [reply]
Yes, a main advantage of requiring that attributes be stored on the graph is how much it would simplify the code for drawing. The disadvantage of course is less flexibility for the user.  It is easy to get lists of attributes from the NetworkX Graph structure.  I believe it would also be easy to put attributes into the graph from lists of values …  but I may be being naive.  If it is indeed true that it is easy in both directions, then we could try rewrite the module to get all attributes from the graph with a preamble that adds the attributes to the graph.  Maybe it wouldn’t help much… but maybe it would simplify the code enough to avoid these kinds of confusions — or at least make debugging them easier?

**Reply by mjschwenne** [reply]
Obviously removing the existing method for graph visualization would require a huge deprecation cycle (target NX 4.0? :laughing: ) and we'd need to make a new system available long before doing anything if we were even interested in doing it. However if Dan's intuition about it simplifying the code for drawing is correct, that may be an avenue to simplify the interface you refer to as "quite complex and arguably fragile" and make it more robust. Even during my short time being involved with NX it seems like issues relating to the graph visualization are relatively frequent so if this possible new interface works well it would reduce user confusion and the developer presence needed to answer issues about it. Hopefully we'll get more people to chime in here about if they're interested, but I'm not exactly holding my breath.
